sdk/eventhub/azure-eventhubs/azure/eventhub/receiver.py

```diff
@@ -331,43 +357,70 @@ def receive(self, max_batch_size=None, timeout=None):
                     event_data = EventData(message=message)
                     self.offset = event_data.offset
                     data_batch.append(event_data)
                 return data_batch
-            except (errors.TokenExpired, errors.AuthenticationException):
-                log.info("Receiver disconnected due to token error. Attempting reconnect.")
-                self.reconnect()
+            except errors.AuthenticationException as auth_error:
+                if connecting_count < 3:
+                    log.info("EventReceiver disconnected due to token error. Attempting reconnect.")
+                    self._reconnect()
+                else:
+                    log.info("EventReceiver authentication failed. Shutting down.")
+                    error = AuthenticationError(str(auth_error), auth_error)
+                    self.close(auth_error)
+                    raise error
             except (errors.LinkDetach, errors.ConnectionClose) as shutdown:
                 if shutdown.action.retry and self.auto_reconnect:
-                    log.info("Receiver detached. Attempting reconnect.")
-                    self.reconnect()
+                    log.info("EventReceiver detached. Attempting reconnect.")
+                    self._reconnect()
                 else:
-                    log.info("Receiver detached. Shutting down.")
-                    error = ConnectError(str(shutdown), shutdown)
+                    log.info("EventReceiver detached. Shutting down.")
+                    error = ConnectionLostError(str(shutdown), shutdown)
                     self.close(exception=error)
                     raise error
             except errors.MessageHandlerError as shutdown:
-                if self.auto_reconnect:
-                    log.info("Receiver detached. Attempting reconnect.")
-                    self.reconnect()
+                if connecting_count < 3:
+                    log.info("EventReceiver detached. Attempting reconnect.")
+                    self._reconnect()
                 else:
-                    log.info("Receiver detached. Shutting down.")
-                    error = ConnectError(str(shutdown), shutdown)
-                    self.close(exception=error)
+                    log.info("EventReceiver detached. Shutting down.")
+                    error = ConnectionLostError(str(shutdown), shutdown)
+                    self.close(error)
                     raise error
+            except errors.AMQPConnectionError as shutdown:
+                if connecting_count < 3:
+                    log.info("EventReceiver connection lost. Attempting reconnect.")
+                    self._reconnect()
+                else:
+                    log.info("EventReceiver connection lost. Shutting down.")
+                    error = ConnectionLostError(str(shutdown), shutdown)
+                    self.close(error)
+                    raise error
+            except compat.TimeoutException as toe:
```

Comment:
I think we can find a better name than "toe" :)

Reply:
Changed to "shutdown", same as other exception handling statements